sphinx_immaterial/cpp_domain_fixes.py

```diff
@@ -0,0 +1,444 @@
+        if self.declaration is None or not isinstance(
+            self.declaration.declaration, ASTMacro
+        ):
+            return
+        args = self.declaration.declaration.args
+        if not args:
+            return
+        for p in args:
+            nn = p.arg
+            if nn is None:
+                continue
+            decl = ASTDeclaration("macroParam", None, p)
+            assert not nn.rooted
+            assert len(nn.names) == 1
+            self._add_symbols(nn, decl, self.docname, self.line)
+
+    CSymbol._add_function_params = _add_function_params
+
+    def get_id(
+        self: ASTMacroParameter, version: int, objectType: str, symbol: CSymbol
+    ) -> str:
+        # the anchor will be our parent
+        return symbol.parent.declaration.get_id(version, prefixed=False)
+
+    ASTMacroParameter.get_id = get_id
+
+    sphinx.domains.c.CDomain.object_types["macroParam"] = sphinx.domains.ObjType(
+        "macro parameter", "identifier", "var", "member", "data"
+    )
+
+
+def _monkey_patch_cpp_expr_role_to_include_c_parent_key():
+    """Includes ``c:parent_key`` in pending_xref nodes created by CPPExprRole.
+
+    This allows CPPExprRole to be used from the context of C object descriptions,
+    in particular macros.
+    """
+    CPPExprRole = sphinx.domains.cpp.CPPExprRole
+
+    orig_run = CPPExprRole.run
+
+    def run(
+        self: CPPExprRole,
+    ) -> Tuple[List[docutils.nodes.Node], List[docutils.nodes.system_message]]:
+        nodes, messages = orig_run(self)
+        c_parent_key = self.env.ref_context.get("c:parent_key")
+        if c_parent_key is not None:
+            for node in nodes:
+                for refnode in node.traverse(condition=sphinx.addnodes.pending_xref):
+                    if refnode.get("refdomain") == "cpp":
+                        refnode["c:parent_key"] = c_parent_key
+        return nodes, messages
+
+    CPPExprRole.run = run
+
+
+def _strip_namespaces_from_signature(
+    node: docutils.nodes.Element, namespaces: List[str]
+):
+    # Collect nodes to remove first, then remove them in reverse order.
+    removals = []
+    for child in node.traverse(condition=sphinx.addnodes.desc_sig_name):
+        parent = child.parent
+        if not isinstance(parent, sphinx.addnodes.pending_xref):
+            continue
+        if (
+            parent["reftype"] != "identifier"
+            or parent["refdomain"] != "cpp"
+            or parent["reftarget"] not in namespaces
+        ):
+            continue
+        grandparent = parent.parent
+        if isinstance(grandparent, sphinx.addnodes.desc_addname):
+            continue
+        index = grandparent.index(parent)
+        if index + 1 >= len(grandparent.children):
+            continue
+        sibling = grandparent[index + 1]
+        if (
+            not isinstance(sibling, sphinx.addnodes.desc_sig_punctuation)
+            or sibling.astext() != "::"
+        ):
+            continue
```

Comment:
Does this account for nested namespaces as well? It's not important to me, but it might be to very "picky" users.

Reply:
No --- would be nice, but not done at the moment.  I added a comment to the documentation.

Reply:
would it be possible (in the meantime) to specify a nested namespace like so:
```py
cpp_strip_namespaces_from_signatures =["abc", "abc::my_ns1"]
```

BTW, I don't see the newly added example stripping the namespace like the admonition describes. It still renders as
![image](https://user-images.githubusercontent.com/14963867/164741410-d6c86f9f-6007-40c4-a3a7-b936575e8805.png)
Does the type need to be documented for the stripping to happen?

Reply:
No it doesn't, since it happens before reference resolution anyway.

My comment was actually wrong.  In fact you are right that nested namespaces are supported --- I revised the example.